exercises/poker/examples/success-standard/src/Poker.hs

```diff
@@ -0,0 +1,51 @@
+module Poker (bestHands) where
+
+import Data.Maybe (fromJust)
+import Data.List  (isPrefixOf, nub, elemIndex, sortBy)
+
+validHand :: String -> Bool
+validHand h = and [ length (words h) == 5
+                  , all (`elem` "A23456789TJQK") (head <$> h')
+                  , all (`elem` "HSDC") (last <$> h')
+                  , all ((==2) . length) h']
+  where
+    h' = parseHand h
+
+parseHand :: String -> [String]
+parseHand = map f . words
+  where
+    f x | "10" `isPrefixOf` x = 'T' : [last x]
```

Comment:
I know there's no test for this currently, but **if** a test were added with a card "10XXXXXH" that would be an invalid card but this code would treat it as valid. In case such a test case were added in the future, then this could would have to be changed. I would like to avoid this by dealing with this now instead of later. Might I suggest something like?

```haskell
    f ('1':'0':x) = 'T' : x 
    f x           = x 
```

Or anything else similar that does the same.

Reply:
But thats, why I created the function ```validHand```
```Haskell
validHand :: String -> Bool
validHand h = and [ length (words h) == 5
                  , all (`elem` "A23456789TJQK") (head <$> h')
                  , all (`elem` "HSDC") (last <$> h')
                  , all ((==2) . length) h']
  where
    h' = parseHand h
```
Or maybe I'm not getting the point here.

Reply:
`parseHand` discards the XXXX portion of `10XXXXH`, doesn't it? So invalid `10XXXXH` will be treated as invalid.

Reply:
parseHand discards the `XXXX` portion of `10XXXXH`, doesn't it? So invalid `10XXXXH` will be incorrectly treated as valid, instead of invalid.

Reply:
Nope, because of ...
```Haskell
length (words h) == 5
````

Reply:
please fix the failing test.

Reply:
`length (words h) == 5` has nothing to do with `10XXXXH`. `10XXXXH` is ONE word.

Reply:
Does this suits your requirements?
 ```Haskell
validHand :: String -> Bool
validHand h = and [ length (words h) == 5
                  , all (`elem` "A23456789TJQK") (head <$> h')
                  , all (`elem` "HSDC") (last <$> h')
                  , all ((==2) . length) h'
                  , all f (words h)]
  where
    h'  = parseHand h
    f x = length x == 2 || (length x == 3 && take 2 x == "10")

parseHand :: String -> [String]
parseHand = map f . words
  where
    f ('1':'0':x) = 'T' : x 
    f x           = x 
```

Reply:
Well, It could be just one of the changes, don't know which one is better

Reply:
You're right that it could just be one of the two changes.

I think the change to `parseHand` (`f ('1':'0':x) = 'T' : x`) is easier to reason about, because it makes fewer assumptions. I can read `parseHand` and say "Okay, this will split the hand into its cards, and replace any 10 with T. Validity can be checked by only examining the output of `parseHand`". In fact, we can cement this if we say that `validHand` should be `[String] -> Bool` instead of `String -> Bool`.

If the change happens in `validHand` (to use `f x = length x == 2 || (length x == 3 && take 2 x == "10")`), then I have to look at both `parseHand` and `validHand` to verify that they together are doing the right thing. I have to say "Okay, `parseHand` split the hand into cards, and deletes any letters after 10, so that means `validHand` needs to look at both the input and output of `parseHand` to be able to do its work." I can't properly understand one without the other.

Reply:
let me think about it a while